push/ups_integration_with_openshift.md

```diff
@@ -0,0 +1,100 @@
+# Unified Push Server Integration with OpenShift 
+
+## Introduction
+
+This proposal outlines some developer experience issues with provisioning, setting up and configuring the Unified Push Server (UPS) on OpenShift, and integrating it with Apps using the Aerogear SDK.
+
+## Problem Description
+
+The Unified Push Server is made up of a Java application & a database (MySQL).
+It also has a dependecy on an auth provider (currently Keycloak) for users of the UPS Administration UI or REST API.
+There are docker images available for UPS.
+There is also a proof of concept UPS Ansible Playbook Bundle (APB) available.
+However, the APB just provisions UPS and MySQL currently.
+It expects an existing Keycloak instance to be running somewhere with the appropriate Realm, Client & Roles already setup.
+
+The second problem is the lack of a consistent login experience when compared to other services (e.g. Jenkins, Prometheus, Grafana).
+As detailed in [Sign On Proposal](../auth/developer-single-sign-on-across-mobile-services.md), the goal is to have a consistent login experience by logging in using OpenShift credentials.
+
+The third problem is the switch in context and terms between creating Mobile Clients & Services in OpenShift, and setting up, configuring & integrating Apps with UPS Push Applications/Variants for Push messages.
+The developer creates a 'Mobile Client' in OpenShift as a logical construct to represent the App they are developing outside of OpenShift.
+In UPS, the developer creates a 'Push Application' and 1 or more 'Variants' depending on the number & type of Apps they are developing outside of Openshift.
+The developer also has to somehow get the Push configuration from UPS and into their App.
+
+The fourth problem is allowing a developer to write custom logic that dynamically sends Push messages to App users.
+This can be done by calling the UPS REST API with the correct creentials/API Key.
+However, configuration is required in UPS to get these credentials.
+These credentials then need to be managed and exposed to the custom code.
+
+All of these problems highlight pain points in the developer experience.
+These problems don't block the developer, but do make it harder than it could be to add Push integration to their Apps. 
+
+## Terms
+
+- TODO
+
+## OAuth provider support 
+
+To address the first & second problem, UPS will be modified to support different auth providers.
+The 2 providers will be Keycloak (to maintain backwards compatibility for existing users) and a generic OAuth Provider.
```

Comment:
> to maintain backwards compatibility for existing users

not just that - also to give the upstream community a chance to continue as normal. Not everybody is ultimately only interested in running the UPS in Openshift. It was designed to run anywhere - not just Openshift.  A tight Openshift integration is the scope of this proposal, as an alternative option to "run it standalone"

Reply:
Great point.
I'll add more info about this

Reply:
oauth-proxy, with no-auth UPS behind